modules/src/applications/transfer/denom.rs

```diff
@@ -0,0 +1,653 @@
+// "transfer/channel-0/transfer/channel-1/uatom" => `["transfer/channel-1", "transfer/channel-0"]`
+// This is done for ease of addition/removal of prefixes.
+#[derive(Clone, Debug, Default, Eq, PartialEq, PartialOrd, Ord, From)]
+pub struct TracePath(Vec<TracePrefix>);
+
+impl TracePath {
+    /// Returns true iff this path starts with the specified prefix
+    pub fn starts_with(&self, prefix: &TracePrefix) -> bool {
+        self.0.last().map(|p| p == prefix).unwrap_or(false)
+    }
+
+    /// Removes the specified prefix from the path if there is a match, otherwise does nothing.
+    pub fn remove_prefix(&mut self, prefix: &TracePrefix) {
+        if self.starts_with(prefix) {
+            self.0.pop();
+        }
+    }
+
+    /// Adds the specified prefix to the path.
+    pub fn add_prefix(&mut self, prefix: TracePrefix) {
+        self.0.push(prefix)
+    }
+
+    /// Returns true if the path is empty and false otherwise.
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+}
+
+impl<'a> TryFrom<Vec<&'a str>> for TracePath {
+    type Error = Error;
+
+    fn try_from(v: Vec<&'a str>) -> Result<Self, Self::Error> {
+        if v.len() % 2 != 0 {
+            return Err(Error::invalid_trace_length(v.len()));
+        }
+
+        let mut trace = vec![];
+        let id_pairs = v.chunks_exact(2).map(|paths| (paths[0], paths[1]));
+        for (pos, (port_id, channel_id)) in id_pairs.rev().enumerate() {
+            let port_id =
+                PortId::from_str(port_id).map_err(|e| Error::invalid_trace_port_id(pos, e))?;
+            let channel_id = ChannelId::from_str(channel_id)
+                .map_err(|e| Error::invalid_trace_channel_id(pos, e))?;
+            trace.push(TracePrefix {
+                port_id,
+                channel_id,
+            });
+        }
+
+        Ok(trace.into())
+    }
+}
+
+impl FromStr for TracePath {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let parts = {
+            let parts: Vec<&str> = s.split('/').collect();
+            if parts.len() == 1 && parts[0].trim().is_empty() {
+                vec![]
+            } else {
+                parts
+            }
+        };
+        parts.try_into()
+    }
+}
+
+impl fmt::Display for TracePath {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let path = self
+            .0
+            .iter()
+            .rev()
+            .map(|prefix| prefix.to_string())
+            .collect::<Vec<String>>()
+            .join("/");
+        write!(f, "{}", path)
+    }
+}
+
+/// Indicates whether the sender chain is acting as a source or sink. Each send to any chain other
+/// than the one it was previously received from is a movement forwards in the token's timeline. In
+/// these instances the sender chain is acting as the source zone.
+pub enum Source {
+    Sender,
+    Receiver,
```

Comment:
We should add another field: `Neither`. If Bob sends Charlie Alice's token, then neither Bob nor Charlie is the source.

Reply:
I don't think that would work, there should ideally only be two cases, when the host chain is acting as a source or sink, might be better to rename the enum and it's  variants to better express this. 
```rust 
     pub enum Zone {
         Source,
         Sink
     }

Reply:
Assuming that we move the `source_chain()` to a new `Ics20Packet` struct as proposed [here](https://github.com/informalsystems/ibc-rs/pull/1989#discussion_r879758017), then returning a `Zone` as you suggest here would force us to add an extra argument to `source_chain()` as so:
```rust
fn source_chain(&self, HostIsReceiving | HostIsSending) -> Zone;
```
This is because a packet has all the information needed to decide whether the sender, receiver or neither is the source of the token being sent (by comparing the denom prefix with the packet's source channel/port), but a packet alone can't tell a given chain if it's source/sink. You need to add the information of if you're sending or receiving the packet. We would then need yet another `enum` for `HostIsReceiving` and `HostIsSending`. I find it more simple to simply have that information directly in the match statements (i.e. return that the source is sender/receiver/neither, and handle what to do in each case differently based on if you're the receiving or sending chain).

In other words, we would still have 2 branches in our match statements. For example, the `match` statement in `process_recv_packet()` would now look like (without including the `Ics20Packet` suggested refactoring):
```rust
match data.token.denom.source_chain(&prefix) {
    Source::Receiver | Source::Neither => { ... },
    Source::Sender => { ... }
}
```
The change here simply properly models the fact that there are 3 cases when asking "who is the source of the tokens in this token transfer" without changing the handling logic. For readability, really.

Does that make sense?

Reply:
Yeah, that makes sense.

Reply:
Will implement using freestanding funcs as discussed.